### ISSUE TITLE
sysdeps/managarm: ioctl will crash rather than EFAULT if given a bad buffer

### ISSUE THREAD

**Issue by ArsenArsen** [managarm/mlibc#376]
Test code: `struct.unpack('HHHH', fcntl.ioctl(0, termios.TIOCGWINSZ, 0))`, working: `struct.unpack('HHHH', fcntl.ioctl(0, termios.TIOCGWINSZ, b"aaaaaaaa"))`

`ioctl` in userspace won't check the buffer address, and hence will not return `EFAULT` and will crash instead.

**Comment by avdgrinten** [comment]
What's the issue here? Are there programs that rely on EFAULT to detect if a memory range is readable/writable?

**Comment by ArsenArsen** [comment]
I haven't ran into that specifically, but simply the behavior is inaccurate, libc syscall wrappers should never cause a program to crash like this

**Comment by Geertiebear** [comment]
I think in general we should factor out the errno translation code, and just provide a general function that does the translation instead of open coding it in every sysdep.

**Comment by ArsenArsen** [comment]
I agree but that would also not fix this, this happens because ioctl is populated in mlibc rather than posix-server

**Comment by Geertiebear** [comment]
How should ``ioctl`` in userspace check if the buffer is valid?

**Comment by ArsenArsen** [comment]
If I knew that I'd have submitted a PR :P

**Comment by avdgrinten** [comment]
I'm not sure if we should make an effort to return EFAULT instead of raising a segfault. Relying on EFAULT sounds very strange.

**Comment by ArsenArsen** [comment]
It's more of a human reliance than a programmatic one, e.g. I discovered this while poking around with ioctls to test SIGWINCH-related changes. I can't think of a very clean way to make a memory access not fail with a synchronous signal, though.

**Comment by Geertiebear** [comment]
Does glibc or musl do this?

**Comment by ArsenArsen** [comment]
the kernel does for both (since linux will just produce an EFAULT, I'm also pretty sure this doesn't happen with the linux sysdeps for that reason, hence specifically mentioning sysdeps/managarm in the subject)
